File: iBurn/src/main/java/com/gaiagps/iburn/database/EventDao.java
package com.gaiagps.iburn.database;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

import io.reactivex.Flowable;

import static com.gaiagps.iburn.database.Event.ALL_DAY;
import static com.gaiagps.iburn.database.Event.CAMP_PLAYA_ID;
import static com.gaiagps.iburn.database.Event.END_TIME;
import static com.gaiagps.iburn.database.Event.START_TIME;
import static com.gaiagps.iburn.database.Event.START_TIME_PRETTY;
import static com.gaiagps.iburn.database.Event.TABLE_NAME;
import static com.gaiagps.iburn.database.Event.TYPE;
import static com.gaiagps.iburn.database.PlayaItem.FAVORITE;
import static com.gaiagps.iburn.database.PlayaItem.ID;
import static com.gaiagps.iburn.database.PlayaItem.LATITUDE;
import static com.gaiagps.iburn.database.PlayaItem.LONGITUDE;
import static com.gaiagps.iburn.database.PlayaItem.NAME;
import static com.gaiagps.iburn.database.PlayaItem.PLAYA_ID;

/**
 * Created by dbro on 6/8/17.
 */

@Dao
public interface EventDao {
    @Query("SELECT * FROM " + TABLE_NAME)
    Flowable<List<Event>> getAll();

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + FAVORITE + " = 1")
    Flowable<List<Event>> getFavorites();

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + FAVORITE + " = 1 AND " + END_TIME + " >= :now")
    Flowable<List<Event>> getNonExpiredFavorites(String now);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + NAME + " LIKE :name GROUP BY " + NAME)
    Flowable<List<Event>> findByName(String name);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + CAMP_PLAYA_ID + " = :campPlayaId GROUP BY " + NAME)
    Flowable<List<Event>> findByCampPlayaId(String campPlayaId);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + PLAYA_ID + " = :playaId AND " + ID + " != :excludingId")
    Flowable<List<Event>> findOtherOccurrences(String playaId, int excludingId);



    //Event-related Queries
    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " +
            START_TIME_PRETTY + " LIKE :day AND "+
            "not(s_time <= :allDayStart AND e_time >= :allDayEnd)"+
            "ORDER BY "
            + ALL_DAY + ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayTimed(String day,String allDayStart,
                                         String allDayEnd);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE (" + START_TIME_PRETTY +
            " LIKE :day AND " + END_TIME + ">= :now AND " +
            "not(s_time <= :allDayStart AND e_time >= :allDayEnd)"+
            " ) ORDER BY "
            + ALL_DAY + ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayNoExpiredTimed(String day,String now,
                                                  String allDayStart,
                                                  String allDayEnd);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE (" + START_TIME_PRETTY +
            " LIKE :day AND "+
            "s_time<= :allDayStart AND e_time >= :allDayEnd"+
            " ) ORDER BY " + ALL_DAY + ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayAllDay(String day,
                                                    String allDayStart,
                                                    String allDayEnd);

      @Query("SELECT * FROM " + TABLE_NAME + " WHERE ("
              + START_TIME_PRETTY + " LIKE :day AND " +
              "not(s_time <= :allDayStart AND e_time >= :allDayEnd) AND "+
              TYPE + " IN (:types)) ORDER BY " + ALL_DAY +
              ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayAndTypeTimed(String day, List<String> types,
                                                String allDayStart,
                                                String allDayEnd);

    @Query("SELECT * FROM " + TABLE_NAME +
            " WHERE (" + START_TIME_PRETTY +
            " LIKE :day AND " +
            END_TIME + ">= :now AND "+
            "not(s_time <= :allDayStart AND e_time >= :allDayEnd) AND "+
             TYPE + " IN (:types)) ORDER BY " + ALL_DAY + ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayAndTypeNoExpiredTimed
            (String day, List<String> types, String now,
             String allDayStart, String allDayEnd);

    @Query("SELECT * FROM " + TABLE_NAME +
            " WHERE (" + START_TIME_PRETTY +
            " LIKE :day AND "
            + TYPE + " IN (:types) AND "+
            "s_time <= :allDayStart AND e_time >= :allDayEnd "+
            ") ORDER BY " + ALL_DAY + ", " + START_TIME + " ASC")
    Flowable<List<Event>> findByDayAndTypeAllDay(String day, List<String> types,
                                                          String allDayStart,
                                                          String allDayEnd);


    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + START_TIME + " BETWEEN :startDate AND :endDate AND " + ALL_DAY + " = 0  ORDER BY " + START_TIME)
    Flowable<List<Event>> findInDateRange(String startDate, String endDate);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE (" + LATITUDE + " BETWEEN :minLat AND :maxLat) AND (" + LONGITUDE + " BETWEEN :minLon AND :maxLon)")
    Flowable<List<Event>> findInRegion(float maxLat, float minLat, float maxLon, float minLon);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + FAVORITE + " = 1 OR ((" + LATITUDE + " BETWEEN :minLat AND :maxLat) AND (" + LONGITUDE + " BETWEEN :minLon AND :maxLon))")
    Flowable<List<Event>> findInRegionOrFavorite(float minLat, float maxLat, float minLon, float maxLon);

    @Insert
    void insert(Event... arts);

    @Update
    void update(Event... arts);
}

File: iBurn/src/main/java/com/gaiagps/iburn/database/EventDao.kt
package com.gaiagps.iburn.database

import androidx.room.Dao
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

/**
 * Created by dbro on 6/8/17.
 */
@Dao
interface EventDao {
    @Query(
        "SELECT " +
            "(SELECT MIN(" + Event.START_TIME + ") FROM " + EventOccurrence.TABLE_NAME + ") " +
            "AS first_start_time, " +
            "(SELECT MAX(" + Event.END_TIME + ") FROM " + EventOccurrence.TABLE_NAME + ") " +
            "AS last_end_time"
    )
    suspend fun getDateRange(): EventDateRange

    @get:Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME
    )
    val all: Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + PlayaItem.PLAYA_ID + " = :id OR e." + Event.EVENT_UID +
            " = :id ORDER BY e." + Event.START_TIME + " LIMIT 1"
    )
    suspend fun getByPlayaId(id: String?): EventWithUserData

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + PlayaItem.ID + " = :id"
    )
    suspend fun getById(id: Int): EventWithUserData

    @get:Query(
        "SELECT e.*, 1 AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e INNER JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " ORDER BY e." + Event.START_TIME
    )
    val favorites: Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, 1 AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e INNER JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.END_TIME + " >= :now ORDER BY e." + Event.START_TIME
    )
    fun getNonExpiredFavorites(now: Long): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + PlayaItem.NAME + " LIKE :name OR e." + PlayaItem.DESC + " LIKE :name GROUP BY e." + PlayaItem.NAME
    )
    fun findByName(name: String?): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " JOIN " + EventFts.TABLE_NAME +
            " ON e." + Event.EVENT_ID + " = " + EventFts.TABLE_NAME + ".rowid" +
            " WHERE " + EventFts.TABLE_NAME + " MATCH :query"
    )
    fun searchFts(query: String?): Flow<List<EventWithUserData>>


    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.CAMP_PLAYA_ID + " = :campPlayaId GROUP BY e." + PlayaItem.NAME
    )
    fun findByCampPlayaId(campPlayaId: String?): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.EVENT_ID + " = :eventId AND e." + PlayaItem.ID + " != :excludingId" +
            " ORDER BY e." + Event.START_TIME
    )
    fun findOtherOccurrences(eventId: Int, excludingId: Int): Flow<List<EventWithUserData>>


    //Event-related Queries
    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.START_TIME + " >= :dayStart AND e." +
            Event.START_TIME + " < :dayEnd" +
            " ORDER BY " + Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findByDay(
        dayStart: Long,
        dayEnd: Long
    ): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE (e." + Event.START_TIME + " >= :dayStart AND e." +
            Event.START_TIME + " < :dayEnd AND e." + Event.END_TIME + " >= :now) ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findByDayNoExpired(
        dayStart: Long,
        dayEnd: Long,
        now: Long
    ): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE (e." + Event.START_TIME + " >= :dayStart AND e." +
            Event.START_TIME + " < :dayEnd AND e." + Event.TYPE + " IN (:types)) ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findByDayAndType(
        dayStart: Long,
        dayEnd: Long,
        types: List<String?>?
    ): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE (e." + Event.START_TIME + " >= :dayStart AND e." +
            Event.START_TIME + " < :dayEnd AND e." + Event.END_TIME +
            " >= :now AND e." + Event.TYPE + " IN (:types)) ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findByDayAndTypeNoExpired(
        dayStart: Long,
        dayEnd: Long,
        types: List<String?>?,
        now: Long
    ): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.START_TIME + " BETWEEN :startDate AND :endDate AND e." + Event.ALL_DAY + " = 0 ORDER BY e." + Event.START_TIME
    )
    fun findInDateRange(startDate: Long, endDate: Long): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE (e." + PlayaItem.LATITUDE + " BETWEEN :minLat AND :maxLat) " +
            "AND (e." + PlayaItem.LONGITUDE + " BETWEEN :minLon AND :maxLon)"
    )
    fun findInRegion(
        maxLat: Float,
        minLat: Float,
        maxLon: Float,
        minLon: Float
    ): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE f." + Favorite.PLAYA_ID + " IS NOT NULL OR ((e." + PlayaItem.LATITUDE +
            " BETWEEN :minLat AND :maxLat) AND (e." + PlayaItem.LONGITUDE +
            " BETWEEN :minLon AND :maxLon))"
    )
    fun findInRegionOrFavorite(
        minLat: Float,
        maxLat: Float,
        minLon: Float,
        maxLon: Float
    ): Flow<List<EventWithUserData>>

    // All days queries (for showing events across all days)
    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " ORDER BY " + Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findAllDays(): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.END_TIME + " >= :now ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findAllDaysNoExpired(now: Long): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.TYPE + " IN (:types) ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findAllDaysAndType(types: List<String?>?): Flow<List<EventWithUserData>>

    @Query(
        "SELECT e.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Event.VIEW_NAME + " e LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON e." + Event.EVENT_UID + " = f." + Favorite.PLAYA_ID +
            " AND e." + Event.START_TIME + " = f." + Favorite.START_TIME +
            " WHERE e." + Event.END_TIME + " >= :now AND e." +
            Event.TYPE + " IN (:types) ORDER BY " +
            Event.ALL_DAY + ", " + Event.START_TIME + " ASC"
    )
    fun findAllDaysAndTypeNoExpired(
        types: List<String?>?,
        now: Long
    ): Flow<List<EventWithUserData>>

}
